### PR TITLE
Do not call each function with $key and $pages.

### LINKED ISSUE #10: Do not call each function with $key and $pages.
From @georgms:
Maybe make `$pages` an instance variable and fill it in the constructor. Then you won't have to pass `$pages` to every method eg. `getName`.

### PR DESCRIPTION
**Pull request by TheKeymaster**
Fix #10

### CODE REVIEW

**Comment by TheKeymaster** [comment]
@georgms could you please review this?

File: DokuwikiXMLExport.php [hunk 1]
```diff
@@ -5,9 +5,8 @@
  * If any bugs occur, please submit a new issue
  * @see https://github.com/findologic/dokuwiki-plugin-findologic-xml-export/issues/new
  */
-if (!defined('DOKU_INC')) {
-    define('DOKU_INC', realpath(dirname(__FILE__) . '/../../../') . '/');
-}
+
+if (!defined('DOKU_INC')) define('DOKU_INC', realpath(dirname(__FILE__) . '/../../') . '/');
```
Comment (georgms) [inline DokuwikiXMLExport.php]: Never use conditionals without blocks (`{ ... }`). At some point you will add another statement and wonder why it does not work.

File: DokuwikiXMLExport.php [hunk 2]
```diff
@@ -50,71 +49,117 @@ class DokuwikiXMLExport
...
+
+    /**
+     * @var array $pages All pageIds.
+     */
+    protected $pages;
+
+    /**
+     * @var int $key Amount of loops.
+     */
+    protected $key;
```
Comment (georgms) [inline DokuwikiXMLExport.php]: `$pages` makes sense as an instance variable, `$key` not so much.

File: DokuwikiXMLExport.php [hunk 3]
```diff
@@ -50,71 +49,117 @@ class DokuwikiXMLExport
...
 
         $items = array();
-        for ($i = $start; $i < $start + $count; $i++) {
-            $item = $exporter->createItem($i);
+        for ($this->key = $start; $this->key < $start + $count; $this->key++) {
+            $item = $exporter->createItem($this->key);
 
             $name = new Name();
-            $name->setValue($this->getName($pages, $i));
+            $name->setValue($this->getName());
```
Comment (georgms) [inline DokuwikiXMLExport.php]: I'd suggest something like `$this->getName($this->pages[$key]`). The `getName()` method should not have to worry about where to find the page, but just about extracting the name from a page. This applies to all similar methods as well.

File: DokuwikiXMLExport.php [hunk 4]
```diff
@@ -123,110 +168,78 @@ public function generateXMLExport($start, $count)
...
         $date->setTimestamp($metadata["date"]["created"]);
         return $date;
     }
 
     /**
+     * Returns the ID / ordernumber.
+     *
+     * @return string Returns the ordernumber.
+     */
+    private function getOrdernumber()
```
Comment (georgms) [inline DokuwikiXMLExport.php]: I would name it `getId` or `getPageId` because it only becomes an ordernumber once it's put there in the export file.

**Review by georgms — CHANGES_REQUESTED** [review]
`$key` should not be an instance variable.

File: DokuwikiXMLExport.php (hunk 1 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Done.

File: DokuwikiXMLExport.php (hunk 2 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Okay. I wasn't sure if this makes sense or not. Thanks!

File: DokuwikiXMLExport.php (hunk 4 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Totally agreed.

File: DokuwikiXMLExport.php (hunk 3 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: This is what you've meant. I was a bit confused about how the behavior should be. But I am totally on your side this makes sense.

**Comment by TheKeymaster** [comment]
@georgms I fixed errything, please check.

File: DokuwikiXMLExport.php [hunk 5]
```diff
@@ -50,33 +50,74 @@ class DokuwikiXMLExport
...
-    public function generateXMLExport($start, $count)
+    private function getPageIds()
     {
         $indexer = new Doku_Indexer();
         $pagesAndDeletedPages = $indexer->getPages();
 
         $excludedPages = $this->splitConfigToArray($this->conf['plugin']['findologicxmlexport']['excludePages']);
-        $pages = null;
+
         foreach ($pagesAndDeletedPages as $page) {
```
Comment (georgms) [inline DokuwikiXMLExport.php]:
```php
$pagesAndDeletedPages = array_filter($pagesAndDeletedPages, function($page, $k) {
  // return true if page description is not empty.
}, ARRAY_FILTER_USE_BOTH);
```

File: DokuwikiXMLExport.php [hunk 6]
```diff
@@ -50,33 +50,74 @@ class DokuwikiXMLExport
...
             if (p_get_metadata($page)['description']) { // Only get pages with content
                 if (!in_array($page, $excludedPages)) { // Exclude pages from config
```
Comment (georgms) [inline DokuwikiXMLExport.php]: Use `array_diff`.

File: DokuwikiXMLExport.php [hunk 7]
```diff
@@ -50,33 +50,74 @@ class DokuwikiXMLExport
...
-        $count = min($total, $count);
+    /**
+     * Formats Config string to an array.
+     *
+     * @param string $config Excluded pages in a string.
+     * @return array Returns the pages that should be excluded as array.
+     */
+    private function splitConfigToArray($config)
+    {
+        $array = explode(',', $config);
```
Comment (georgms) [inline DokuwikiXMLExport.php]: Maybe use `preg_split`.

File: DokuwikiXMLExport.php [hunk 8]
```diff
@@ -88,33 +129,33 @@ public function generateXMLExport($start, $count)
             $item = $exporter->createItem($i);
```
Comment (georgms) [inline DokuwikiXMLExport.php]: Use `array_slice` and `foreach`.

**Review by georgms — CHANGES_REQUESTED** [review]
Some minor changes.

File: DokuwikiXMLExport.php (hunk 5 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Done.

File: DokuwikiXMLExport.php (hunk 6 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Done.

File: DokuwikiXMLExport.php (hunk 7 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Done

File: DokuwikiXMLExport.php (hunk 8 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Done

**Comment by TheKeymaster** [comment]
@georgms I fixed errything, please check :))))

File: DokuwikiXMLExport.php [hunk 9]
```diff
@@ -79,17 +79,17 @@ private function getPageIds()
...
-        $excludedPages = $this->splitConfigToArray($this->conf['plugin']['findologicxmlexport']['excludePages']);
-
-        foreach ($pagesAndDeletedPages as $page) {
-            if (p_get_metadata($page)['description']) { // Only get pages with content
-                if (!in_array($page, $excludedPages)) { // Exclude pages from config
-                    $ids[] = $page;
-                }
+        // Get all pages that do have a description
+        $pagesAndDeletedPages = array_filter($pagesAndDeletedPages, function ($page, $k) {
+            if (p_get_metadata($page)['description']) {
```
Comment (georgms) [inline DokuwikiXMLExport.php]: You can make it even shorter by just using the expression as a return value, something like `return (p_get_metadata($page)['description'] !== '');`

File: DokuwikiXMLExport.php [hunk 10]
```diff
@@ -100,17 +100,15 @@ private function getPageIds()
      */
     private function splitConfigToArray($config)
     {
-        $array = explode(',', $config);
-        $trimmedArray = array_map('trim', $array);
-        return $trimmedArray;
+        return preg_split('(\s*,\s*)', $config);
```
Comment (georgms) [inline DokuwikiXMLExport.php]: The parentheses `(...)` are used for capturing group when doing regex matching. They serve no purpose when splitting.

**Review by georgms — CHANGES_REQUESTED** [review]
Two minor issues, looks pretty nice.

File: DokuwikiXMLExport.php (hunk 9 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Yep.

File: DokuwikiXMLExport.php (hunk 10 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Done

**Comment by TheKeymaster** [comment]
@georgms Fixed!!! :tada: :tada:

**Review by georgms — APPROVED** [review]

**Comment by TheKeymaster** [comment]
@georgms Will merge.

**Pull request merged**